signer/src/util.rs

```diff
@@ -0,0 +1,290 @@
+
+use std::{
+    cmp::min,
+    future::Future,
+    ops::Deref,
+    sync::{
+        atomic::{AtomicU8, AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+use thiserror::Error;
+
+use crate::error::Error;
+
+const DEFAULT_MINIMUM_RETRY_COUNT: usize = 3;
+
+/// Error variants for the fallback client.
+#[derive(Debug, Error)]
+pub enum FallbackClientError {
+    /// All failover clients failed within the retry limit
+    #[error(
+        "all fallback clients failed to execute the request within the allotted number of retries"
+    )]
+    AllClientsFailed,
+
+    /// No endpoints were provided
+    #[error("no endpoints were provided")]
+    NoEndpoints,
+
+    /// Retry count must be greater than zero
+    #[error("retry count must be greater than zero")]
+    RetryCountZero,
+}
+
+/// A fallback-wrapper that can failover to other clients if the current client fails.
+pub struct ApiFallbackClient<T> {
+    inner: Arc<InnerApiFallbackClient<T>>,
+}
+
+impl<T> Deref for ApiFallbackClient<T> {
+    type Target = InnerApiFallbackClient<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T> Clone for ApiFallbackClient<T> {
+    fn clone(&self) -> Self {
+        Self { inner: Arc::clone(&self.inner) }
+    }
+}
+
+/// Inner implementation of the fallback client.
+pub struct InnerApiFallbackClient<T> {
+    inner_clients: Vec<T>,
+    last_client_index: AtomicUsize,
+    retry_count: AtomicU8,
+}
+
+impl<T> InnerApiFallbackClient<T> {
+    /// Set the number of retries to perform before giving up.
+    pub fn set_retry_count(&self, retry_count: u8) -> Result<(), FallbackClientError> {
+        if retry_count == 0 {
+            return Err(FallbackClientError::RetryCountZero);
+        }
+
+        self.retry_count.store(retry_count, Ordering::Relaxed);
+
+        Ok(())
+    }
+
+    /// Get a reference to the current inner API client.
+    pub fn get_client(&self) -> &T {
+        &self.inner_clients[self.last_client_index.load(Ordering::Relaxed)]
+    }
+
+    /// Execute a closure on the current client, falling back to remaining clients
+    /// if the closure returns an error.
+    pub async fn exec<'a, R, E, F>(&'a self, f: impl Fn(&'a T) -> F) -> Result<R, Error>
+    where
+        E: std::error::Error + std::fmt::Debug,
+        E: Into<Error>,
+        F: Future<Output = Result<R, E>> + 'a,
+    {
+        let retry_count = self.retry_count.load(Ordering::Relaxed);
+        for i in 0..retry_count {
+            let client_index = self.last_client_index.load(Ordering::Relaxed);
+            let result = f(&self.inner_clients[client_index]).await;
+
+            if let Err(error) = result {
+                tracing::warn!(%error, retry_num=i, max_retries=retry_count, "failover client call failed");
+                self.last_client_index.store(
+                    (client_index + 1) % self.inner_clients.len(),
+                    Ordering::Relaxed,
+                );
+                continue;
```

Comment:
Yeah this is just a naive "something that works for right now", I fully expect that it will be iterated upon :)

You can still achieve the same effect as `retry_if` though even with this, you just bake your evaluation of `f` within `exec()` and return an `Ok` if you don't want to retry.